lib/remote_node.go

```
@@ -0,0 +1,468 @@
+		rn.Disconnect()
+		return
+	}
+
+	vMeta := rn.GetHandshakeMetadata()
+	if vMeta.versionTimeExpected != nil && vMeta.versionTimeExpected.Before(time.Now()) {
+		glog.V(1).Infof("RemoteNode.HandleVersionMessage: Requesting PeerDisconnect for id: (%v) "+
+			"version timeout. Time expected: %v, now: %v", rn.peer.ID, vMeta.versionTimeExpected.UnixMicro(), time.Now().UnixMicro())
+		rn.Disconnect()
+		return
+	}
+
+	// Save the version nonce so we can include it in our verack message.
+	vMeta.versionNoncesReceived = verMsg.Nonce
+
+	// Set the peer info-related fields.
+	vMeta.userAgent = verMsg.UserAgent
+	vMeta.ServiceFlag = verMsg.Services
+	vMeta.advertisedProtocolVersion = NewProtocolVersionType(verMsg.Version)
+	negotiatedVersion := rn.params.ProtocolVersion
+	if verMsg.Version < rn.params.ProtocolVersion.ToUint64() {
+		negotiatedVersion = NewProtocolVersionType(verMsg.Version)
+	}
+	vMeta.negotiatedProtocolVersion = negotiatedVersion
+	vMeta.minTxFeeRateNanosPerKB = verMsg.MinFeeRateNanosPerKB
+	timeConnected := time.Unix(verMsg.TstampSecs, 0)
+	vMeta.timeConnected = &timeConnected
+	vMeta.timeOffsetSecs = verMsg.TstampSecs - time.Now().Unix()
+	vMeta.StartingBlockHeight = verMsg.StartBlockHeight
+
+	// Update the timeSource now that we've gotten a version message from the peer.
+	rn.cmgr.AddTimeSample(rn.peer.Address(), timeConnected)
+
+	if !rn.peer.IsOutbound() {
+		// Respond to the version message if this is an inbound peer.
+		rn.sendVersionMessage(responseNonce)
+	}
+	// After sending and receiving a compatible version, complete the
+	// negotiation by sending and receiving a verack message.
+	verackTimeExpected := time.Now().Add(rn.params.VersionNegotiationTimeout)
+	vMeta.verackTimeExpected = &verackTimeExpected
+	if err := rn.sendVerack(); err != nil {
+		glog.Errorf("RemoteNode.HandleVersionMessage: Problem sending verack message to peer (id= %d): %v", rn.peer.ID, err)
```

I generally prefer to return errors and require the caller to handle them instead of glogging them. It just becomes tricky to track down bugs later on.